### PR TITLE
Fix tool change priming

### PR DESCRIPTION
**Pull request by RFBomb**
### Description

Fixed the tool change unretract logic when using `TOOLCHANGE_FILAMENT_SWAP` but not  `TOOLCHANGE_FS_PRIME_FIRST_USED`

### Requirements

More than 1 Extruder Configured
TOOLCHANGE_FILAMENT_SWAP

### Benefits

Previously:
if not using `TOOLCHANGE_FS_PRIME_FIRST_USED` (disabled by default), then the unretract would simply never occur, defeating the entire purpose of enabling the filament swap feature.

Now:
After the 'Prime' and the 'Extra Prime' movements occur, log that the selected extruder has been primed. If you switch extruders again, then this will allow that extruder to retract. If it hasn't been primed yet, then it will not retract.

### Related Issues

Didn't see any related issues created by other people. But I noticed the problem when testing out the TOOLCHANGE_FILAMENT_SWAP feature. I saw that it would just repeatedly prime the nozzle, but would never retract like the description says it should

### CODE REVIEW

**Comment by thinkyhead** [comment]
It would be good to clarify the teams and the meanings of these options.
- …`TOOLCHANGE_FILAMENT_SWAP` is described as "Retract and prime filament on tool-change to reduce ooze and stringing and to get cleaner transitions." Users may not know where this priming will occur, where the primed filament will go, and other details. Users may want to know how a "filament swap" differs from doing a "nozzle wipe" or "nozzle clean" or even a "nozzle park" since all of these procedures involve movement, retraction, priming, etc., and are meant to aid in more complex printing setups.
- …`INIT_BEFORE_SWAP` is now described as "Swap uninitialized extruder" and this is not quite clear. What makes an extruder "uninitialized" and what is being "swapped"? It would be good to lay out the reasons why this option should or should not be enabled and the effects that should be expected.
- …`PRIME_FIRST_USED` is now described as "Prime on the first T0 (If other, `TOOLCHANGE_FS_INIT_BEFORE_SWAP` applied)." This should probably include an explanation of why you would or would not want to prime the first time the machine sees a `T0` command. The rest of the description lightly implies that connecting to a host will cause a `T0` to be run, either because the host is sending it or because Marlin itself is activating its first tool. It would help to have a little more explanation of the consequences of using or not using this option.

In general these kinds of options may have some weakness, since they expect a certain behavior from host software which might change in the future. These options give the impression of being workarounds for a slightly broken system of slicers / hosts / firmware handshaking, so the more details we can give about the weird behaviors these are meant to work around, the easier it will be to clean them up in the future.

**Comment by RFBomb** [comment]
I agree that it is confusing. I was pretty surprised myself that when I enabled the feature, it just straight up didn't do what it said it would (retract). That is why i went digging and found the issue. The code that I had removed in my original PR, frankly I'm not sure what issue that code was trying to resolve, other than an obscure way of disabling the retraction. If that was the goal, then it would've been better just to have another define for `TOOLCHANGE_FS_UNRETRACT_LENGTH`. 

Also, after looking at it `TOOLCHANGE_FS_EXTRA_RESUME_LENGTH` is another confusing term. It is described as 'extra length for better restart'. I would interpret this as meaning that the initial prime distance is the length + this. THEN there is the `...EXTRA_PRIME` distance, which is at a slower speed if you want it. 

But looking through the code, `...RESUME_LENGTH` is only ever to apparently recover from `...WIPE_RETRACT `. So the process is:
- Retract current nozzle (now)
- Raise, Park, Select New Nozzle
- Prime + Extra Prime
- Retraction for Wipe_Retract length
- Move to destination
- Prime again using ( Resume_Length + Wipe_Retract )  as the length.

Its very confusing, as this process isn't really described in the notes, and the order in the config file makes it appear as though 
`TOOLCHANGE_FS_EXTRA_RESUME_LENGTH` is added onto `TOOLCHANGE_FS_LENGTH`, when it is not, atleast not directly. 


`TOOLCHANGE_FS_INIT_BEFORE_SWAP` i straight up didn't understand what that was meant for.  As far as the rest of the response, I interpreted 'unitialized extruder' as meaning it had never been selected (primed), so it needs to prime the first time it is selected. But, I don't know how this differs from simply having `TOOLCHANGE_FILAMENT_SWAP` enabled, since at the end of the tool switch it will then prime the nozzle according to the `...FS_LENGTH`. As you said, its confusing. 
It wasn't until writing this response that I realize I think it means using the 'Extra Priming Feedrate' `TOOLCHANGE_FS_PRIME_SPEED ` for the entire length of the first 'prime' action if this is the first time selecting the extruder since power-on. 

`TOOLCHANGE_FS_PRIME_FIRST_USED` I took as it simply primes T0 when the printer T0 is issued for the first time. I am unsure why this was really even needed, since when you swap over to a new extruder it runs the prime length anyway. Unless this is possibly a workaround to have it basically do a double-prime action when first selected. But even so, I can't tell if this will perform this action on T1 when swapping over to it, or if it ONLY affects T0. And thats after evaluating the code, which 99% of users will never even attempt.

File: Marlin/src/module/tool_change.cpp [hunk 1]
```diff
@@ -93,7 +93,6 @@
 
 #if ENABLED(TOOLCHANGE_FILAMENT_SWAP)
   #include "../gcode/gcode.h"
-  bool extruder_primed[EXTRUDERS];  // Log that extruders have been primed
```
Comment (RFBomb) [inline Marlin/src/module/tool_change.cpp]: This follows the same structure as the example on line 48/49, building an array to store the true/false data in for the extruders. 
```
#if ENABLED(TOOLCHANGE_FS_INIT_BEFORE_SWAP)
  bool toolchange_extruder_ready[EXTRUDERS];
#endif
```

What advantage does the replacement on line 962 `static uint8_t extruder_primed = 0; // Flags for primed extruders` have over the array I coded in?

**Comment by thinkyhead** [comment]
What would happen if these options were entirely removed? Then, starting from a clean slate it could be redesigned in a way that is more minimal and more sensible. I feel the original author did rush this feature and tailored it too much to a single use case, and it is a bit scattershot. I would dive into a complete rewrite myself but I don't have a machine suitable for testing. Maybe all that is needed is some rearrangement and a snip here and there….

We ought to start with deciding on some terminology so it's clear where in the tool change procedure each setting applies. The word "swap" is used in Marlin typically to refer to the first retract or purge after a tool has been changed, thus sitting idle, oozing, and losing pressure for some period of time. This is to distinguish from a "non-swap" situation, so… I guess that makes it a retronym.

"Retract" is not a problematic term. We use "unretract" and "prime" sort of interchangeably, but if we're being more exact, an "unretract" is the distance moved forward to balance an earlier "retract" and a "prime" is an extra move forward to give extra pressure. The word "purge" implies a longer extrusion to clear out old material.

To deal with the immediate issue of bad priming behavior, let's make sure this PR fixes at least that much, and please apply whatever changes you think will help make it simpler and easier to understand.

**Comment by goroph1** [comment]
One of the big struggles I was running into with my Cyclopes clone print head, is that it would retract the filament before lifting the nozzle, this would leave defect marks everywhere a tool change occurred. I have my system set up to purge the Nozzle off the print bed, and I really prefer to have the entire process taken care of in firmware. My c++ abilities are much better then my Gcode abilities. I think better commenting would help, but I did get it working with good results by rearranging the code a bit.

[tool_change.cpp.txt](https://github.com/MarlinFirmware/Marlin/files/6087410/tool_change.cpp.txt)

The comments I added to the file are more to highlight the changes I have made than anything else,

**Comment by RFBomb** [comment]
I would be happy to assist in working on the rewrite as well. Since I
already have a fork started for this PR, the work could be done there.

We should outline what we want to do, provide configurable order as goroph1
mentions, and some basic other settings.

I think the structure is mostly there, but it could be cleaned up for sure.
I noticed the ‘first prime’ routine was essentially a huge copy-paste of
the rest of the routine for example.

**Comment by thinkyhead** [comment]
I'll have a look at the `tool_change.cpp` file (posted above) later today and we can compare notes.

**Comment by RFBomb** [comment]
ok, I've been looking at it, and I think that the function `void tool_change_prime()` is unnecessary. 
This function is essentially a duplicate of the main `tool_change` function. The main difference, from what I can tell, is that it will basically prime T0 it twice the first time T0 is called, without affecting any other nozzle. I think it can be safely removed. 

Using it also appears to cause the printer to move to nozzle park, prime, move to print, move to nozzle park, prime again, then move back to print. This is because after the initial prime, it just continued on with the remainder of the `tool_change` function. 

I propose creating a function for `void extruder_prime` that does all the 'prime' extruder movements & checks `too_cold`. This way, if the 'double prime' I refer to above was the intention of that `tool_change_prime` routine, it can be simplified by just calling the `extruder_prime` routine twice if needed.

edit: Done more evaluation. 
The `void tool_change_prime()` function will prime only when: `enable_first_prime && !first_tool_is_primed && new_tool == old_tool) `. I understand this to mean that it will only every prime T0  when there is multiple extruders enabled (otherwise it exits the tool_change function prior to running this),  and the gcode command 'T0' is issued. It also only does this once per power-up. 

The main advantage / difference to this function when compared to the rest of the tool_change routine is that it will force T0 to move to the ToolPark position (if enabled), prime, then return back to where it was previously. What I referenced about 'being primed twice' was false, since `old_tool == new_tool`, which prevents priming a tool thats already selected. I still think that this function can be eliminated. In its place, make the following modifications:

Current code that checks if filament-swap is required (line 944 of tool_change.cpp) :
`if (new_tool != old_tool || TERN0(PARKING_EXTRUDER, extruder_parked))`

Updated:
- `if (new_tool != old_tool || ( new_tool == old_tool && !TEST(extruder_did_first_prime, new_tool)) ) || TERN0(PARKING_EXTRUDER, extruder_parked))`
- Remove lines 934-942 (logic to decide to call `tool_change_prime`)
- Remove lines 791-854 (tool_change_prime routine)

**Comment by thinkyhead** [comment]
Yes, those changes seem very sensible! You can add commits to your `Update_ToolChange_Retraction` branch and push to your (origin) remote, and it will directly update this PR.

**Comment by RFBomb** [comment]
I've been toying around with it when I've had the chance, but haven't actually tested anything yet. The push I just performed is a heavy WIP. I uploaded it as a seperate file since I'm working side-by-side with the original one to ensure the sequence stays basically the same. 

Unsure when I'll get a chance to actually test it out, but I'm hoping in the next week or so I'll have 'finished' the preliminary and be able to do back to back testing to verify nothing quirky happens when switching tool heads

**Comment by thinkyhead** [comment]
Merged and resolved conflicts. Prevented compilation of the extra toolchange work in progress file.

**Comment by thinkyhead** [comment]
This looks about ready, but should be re-tested. And, we need to remove `tool_change_Revised.cpp` — so if any part of that is still being considered for inclusion, now is the time to port it over. I'll point testers towards this PR, but there are not a lot of tool-changers out there.

**Comment by RFBomb** [comment]
My tool change setup should be able to test this, as I had found the issue initially. 

I can't even recall how I found it, since I've had other code projects I've been dealing with since creating this thread. 

The 'revised' file should be fine to delete, as I believe I was using that more to compare & showcase a proposed fix (compared using Notepad++ as opposed to git changes, since when I created this I was still familiarizing myself with git/Visual Studio functionality.)

About 2 months ago or so I couldn't recall what changes I had done to various code files and saw some may have already been merged, so I had just resorted to using the main branch here for my printer.

**Comment by RFBomb** [comment]
Before I test this, could you pull in the other changes from the bug fix branch (if you haven't already done so) - I know this a lot of commits have been done since this PR was initially created

**Comment by thinkyhead** [comment]
I did a git merge from `upstream/bugfix-2.0.x` just now and there were no conflicts. But, GitHub is being glitchy and preventing me from pushing the merge. So, you can go ahead and do that any time.

**Comment by RFBomb** [comment]
Awesome to hear. I am traveling through next week, so can't test it yet but will let you know. 

Is there anything specific you'd like me to try and test for? If needed, I'll write out the gcode accordingly. 

Testing will be done on an ender 3 that has been modified to have 2 extruders, 24mm X offset between them.

**Comment by RFBomb** [comment]
Test Setup
-

I used M503 to echo back my settings, and I received this:
```
Tool-changing:
M217 S16.00 B0.00 E0.00 P276.00 R3000.00 U1500.00 F255 D10 Z2.00
```

Notably, the `D` parameter is missing from this page: https://marlinfw.org/docs/gcode/M217.html, but I think that is the `TOOLCHANGE_FS_FAN_TIME`, since thats the only thing in my config that has a value of 10 here.

Attached are my config files (note this is a SKR1.3 board that also has a BL-Touch and some other customization (firing relays for the fan enable has been set up for example)
[Ender3_DualExtruder.zip](https://github.com/MarlinFirmware/Marlin/files/8346552/Ender3_DualExtruder.zip)

My testing thus far has been manually sending the printer commands via OctoPrint and just watching the printer react. 
All tests have been performed after a simple G28, which lands the T0 about 5mm above bed center.
I have also set my offsets to -55 for both X, -12.7 for Y, and -2 for Z. This allows me to exaggerate the movement and use my BL-Touch for my 'touchdown' for nozzle 2 instead of the actual nozzle 2 that has a Y in line with nozzle 1.

Printer Crash:
-

- I crashed my printer while adjusting the tool offsets using my Ender's stock display. 
   - I was testing simple T0/T1 commands prior to deciding to update the values. 
   - I had last used the T1 command.
   - I then adjusted the 'Nozzle 2 Offset' from -50 to -55.
   - This caused my printer to automatically perform a G28.
   - In doing so, it hit the X-0 and Y-0 sensor, then began to 'spaz' out as it continued to attempt to drive into that, until a few seconds later where it errored out and requested I power cycle it.
   - At this time, I'm not sure why it crashed like it did, It had done successfull G28's a few minutes prior to this.
   - I attempted to repeat it (unsuccessfully). 
       - Modifying these settings after a G28 while T0 is active: Printer does not move.
       - Swapped to T1, modified settings, printer did not move. 
       - ~I have no idea what made it perform a G28 when modifying those settings before~, but I witnessed it happen three times in total. Of those three times, only one had crashed/errored.
             - Further testing i was able to replicate it after performing multiple T0/T1 commands THEN changing the offsets. Extruder T1 had to move to trigger this G28.
       - Multiple attempts to replicate this failed, and all tests of G28 (I'm using Quick home where it does X and Y at same time) worked as expected.
   - Below is what Octoprint received during the crash (Temperature and 'Wait/Busy' have been suppressed)

 ```
Recv: echo:Active Extruder: 0
Recv: ok P63 B31
[...]
Send: T1
[...]
Recv: echo:Active Extruder: 1
Recv: ok P63 B31
[...]
Recv: echo:Active Extruder: 0
[...]
Recv: echo:Homing Failed
Recv: Error:Printer halted. kill() called!
Changing monitoring state from "Operational" to "Error"
```

Tool Change Testing
-

During my testing, which at this stage mostly involved just using T0 & T1 commands back to back, I noted a few things:
- When moving from T0 to T1, 
    - T0 retracts a bit
    - T1 Extrudes then immediately retracts
    - Printer moves to bring T1 to desired location. (Where T0 previously was)
    - T1 now primes itself.
 
https://user-images.githubusercontent.com/20431767/160040545-a645a5ef-4378-4bfd-8881-042d5f6e2591.mp4

- When moving from T1 to T10 
    - T1 retracts a bit
    - T0 Extrudes then immediately retracts
    - Printer moves to bring T0 to desired location. (Where T1 previously was)
    - T0 now primes itself

https://user-images.githubusercontent.com/20431767/160040372-0171feb6-7caf-438d-addf-b6f27c5f433f.mp4


My conclusions as of this build: 
For ease of use, I will call the 'Active' extruder 'OLD' and extruder being selected 'NEW' here.

1. OLD performs its retract distance correctly. 
2. NEW then extrudes a bit
    - Unsure why this occurs, but it likely shouldn't. Maybe have to do with 'Wipe' setting?
3. NEW then retracts a lot.
    - This is likely meant to be the OLD extruder instead, as it appears to be retracting it the 'swap' distance.
4. Printer then relocates according to offsets for NEW.
5. NEW then primes itself 
    - This is currently always priming what appears to be the 'swap' distance, as expected. 

So I believe steps 2 and 3 are incorrect.

**Comment by RFBomb** [comment]
These last commits I did seem to be working much better.
Purging seems much more consistent to me, and modifying it will be easier in the future due to isolated method that will run whenever a purge is used. 

Cutting is working well, but I still have to test Cutting recovery. Extra prime distance works well too. 

Then theres obviously the checks that failed, they all seem to be reporting the same error: "Unknown development platform"
example:
```
[Test BIGTREE_BTT002 --target clean] Setup Build Environment...
Error: Unknown development platform 'ststm32@~12.1'
Failed!
```

no idea what thats about

File: Marlin/src/module/tool_change.cpp [hunk 2]
```diff
@@ -1140,13 +1170,16 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
       TEMPORARY_BED_LEVELING_STATE(false);
     #endif
 
+    // TODO: This should maybe be expanded to take single nozzle into account, as well as other scenarios. 
+    // Currently, it will only prime the default extruder when that T# is called and that T# is already selected.
+    // Ex: Printer boots up, T0 is default. Terminal sends T1 -> This does nothing (T0 does not retract since it has not been primed), swaps to T1 -> That may be problematic in single nozzle if we don't know which was primed last.
+    // Ex: Printer boots up, T0 is default. Terminal sends T0 -> This primes T0.
     // First tool priming. To prime again, reboot the machine.
     #if ENABLED(TOOLCHANGE_FS_PRIME_FIRST_USED)
```
Comment (RFBomb) [inline Marlin/src/module/tool_change.cpp:1166]: @thinkyhead 
I added comments to this based on what I can tell it does. I think this needs to be reviewed prior to merging in. 

I think that its likely fine to keep the `TOOLCHANGE_FS_PRIME_FIRST_USED` feature, as long it has comments saying that when a T0 is issued for the first time while T0 is selected (so on printer startup), it will prime T0 for the entire priming distance. 

But what I think needs to be reviewed here is that we should likely also do this for SingleNozzles, or use the `TOOLCHANGE_FS_INIT_BEFORE_SWAP` flag here. It would initialize T0 prior to swapping to T1, ensuring that T0's filament ran forward some distance instead of potentially backing out completely. 

That said, the other change I did below would prevent that backing on first change since the tool itself was never 'primed'.  (which thats its own issue here I'll admit. But short of assuming it was primed at the end of the ToolChange() method, I don't see a way to resolve that. But since that assumption would land us on the side of caution, it may be the better thing to do than not assume.)

File: Marlin/src/module/tool_change.cpp [hunk 3]
```diff
@@ -1181,6 +1214,8 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
           else if (TEST(extruder_was_primed, old_tool)) 
           {
             // Retract the old extruder if it was previously primed
+            // To-Do: Should SingleNozzle always retract?
```
Comment (RFBomb) [inline Marlin/src/module/tool_change.cpp:1202]: This location is where the old extruder retracts when swapping to the new extruder. 
Currently, as stated in the other review comment, it only retracts if it was known to have primed via T0/T1T#. 
This is where we may want to always retract the old nozzle for SINGLENOZZLE setups. Unsure - looking for your thoughts @thinkyhead

**Comment by RFBomb** [comment]
alright, I've tested the filament swap portion of it, and added debug echo backs to verify the functionality.

I added a couple code review comments I'd like input on to see if those logic sequences should be updated. But from what I can tell everything is functional as expected

File: Marlin/src/module/tool_change.h [hunk 4]
```diff
@@ -24,20 +24,19 @@
...
     #if ENABLED(TOOLCHANGE_FILAMENT_SWAP)
-      float swap_length;            // M217 S
-      float extra_prime;            // M217 E
-      float extra_resume;           // M217 B
-      int16_t prime_speed;          // M217 P
-      int16_t retract_speed;        // M217 R
-      int16_t unretract_speed;      // M217 U
-      int16_t fan_speed;            // M217 F
-      int16_t fan_time;             // M217 D
+      float swap_length,            // M217 S
```
Comment (RFBomb) [inline Marlin/src/module/tool_change.h]: @thinkyhead 
So this change, while it does look cleaner, the reason I had put the declarations on each line item was so that VisualStudio would show the relevant comment for the particular settings when hovering them in the code. 
Since functionally code-wise they are identical, I think its better to have them explicitly declared for easier time when working with them in the code.

image

versus:

image

**Comment by RFBomb** [comment]
Testing 2 extruders with 2 nozzles seems to work great, everything as expected. 

SINGLENOZZLE - Works as expected.
-
- Seemed to work just fine for using T1 then T0 then T1 again
- Note: Using T1 did NOT retract T0 on first use, since T0 had not been recorded as being primed yet.
    - Recommend using `TOOLCHANGE_FS_PRIME_FIRST_USED` and `TOOLCHANGE_FS_SLOW_FIRST_PRIME` with SINGLENOZZLE for this reason.


TOOLCHANGE_FS_PRIME_FIRST_USED  -- Looks Good!
-
- Previously, this would only issue the priming command if ExtraPrime had a value > 0. This was not made clear to any users, so I eliminated that check to perform the prime if this is enabled.
    - Really, this will only affect T0, and it will only affect it the first time its selected on power up. 
- With `M217 V0`, T0 did NOT do its first init on printer power on when issuing a T0 as the first command after G28.
    - This matches the description from Config_adv.
    - Therefore, issuing a T1 loaded filament with T1 without having retracted T0. 
    - I can see this being problematic potentially for SingleNozzle scenarios, but I think that it will just be something they have to be aware of. 
    - Alternatively, we can put another toggle to always unload the old tool, even if it was not previously primed. But this risks backing out the filament completely. As such, I don't think this is a good idea. Plus, we have no idea what was last loaded filament, so we are back to user being aware of their printer's state, which they should be anyway when starting a print.
- With `M217 V0`, T0 DID do its first init on printer power on when issuing a T0 as the first command after G28.
    - Once again, this is in line with the description.
    - Issuing a T1 then did unload T0. 

TOOLCHANGE_FS_SLOW_FIRST_PRIME
-
- `TOOLCHANGE_FS_INIT_BEFORE_SWAP` renamed to `TOOLCHANGE_FS_SLOW_FIRST_PRIME`
- `TOOLCHANGE_FS_INIT_BEFORE_SWAP` did not clearly communicate what the setting was doing. As such, it has been renamed to `TOOLCHANGE_FS_SLOW_FIRST_PRIME`.
- I have already put this into the SanityCheck and done a search against the repo, changing it in the few places I saw it.

**Comment by RFBomb** [comment]
@goroph1 
I took a peek at the file you uploaded, honestly I have no idea what you changed to correct your issue. 
From what I can tell, when comparing it to the current 'live' version of the file on the bugfix2.0 repo, is that it appears somewhat old.
Notably, it is missing this define completely: `TOOLCHANGE_ZRAISE_BEFORE_RETRACT`, which resolves the exact issue you describe. 

As such, my request to you would be to try out this PR's code and report back. I would pull this fork into its own folder, drop in your configs, update them as needed, enable `TOOLCHANGE_ZRAISE_BEFORE_RETRACT` in config_adv, and build it.

File: Marlin/src/module/tool_change.cpp (hunk 2 above)
Comment (RFBomb) [reply Marlin/src/module/tool_change.cpp:1166]: Following up thoughts:
The original comment I made here was followed with some testing and several commits. As I stated in the main thread for this PR, I think this functionality is fine, as it is now well-documented in the config_adv file where it is set up. That interaction is just something the user will have to be aware of, and it also will only occur on first power up. I would mark this as closed, but I can only delete it. ( I didn't opt to delete due to wanting to preserve my thoughts on the subject)

TLDR: This 'review' comment can be ignored as its working fine.

File: Marlin/src/module/tool_change.cpp (hunk 3 above)
Comment (RFBomb) [reply Marlin/src/module/tool_change.cpp:1202]: Followup: I think this is fine, no change needed. I left the 'To-Do' comment here in case it wants to be looked at in the future, though I don't think it will be needed.

**Comment by RFBomb** [comment]
@thinkyhead 
I left several review comments on the 'Files Changed' tab for the Tool_Change.cpp and Tool_Change.h files. 
- The two on Tool_Change.cpp can be closed out I think. I left them up to keep my thought process preserved, but left followup comments on them with reasoning for saying its fine as is based on changes I've made. 
- The one for the `.h` file though I'd like you to actually review and respond here, as I believe I bring up a good point in it and just want your thoughts on the _why_ you reverted my change

Other than that, as far as I can tell everything is working well and this should be 'mergable'. 
- The only other thing that I can think of potentially doing is that the 'Tool Change Park' positions are used multiple times in the code. Each time they are used, its the same several lines copy-pasted into the code sections, each of them wrapped in an `#if Tool_Change_park` compiler statement. 
    - These could be put into a single method since they are currently done the same way in those locations.
    - I didn't touch these currently due to potentially wanting to modify them individually later on, since these do involve moving XYZ. So while I think it _could_ be compacted into a method that is referenced, I don't know if theres a benefit to do so.

**Comment by thinkyhead** [comment]
> I left several review comments on the 'Files Changed' tab for the Tool_Change.cpp and Tool_Change.h files.

Your comments aren't visible under the Files tab here. Did you put them someplace else?

**Comment by RFBomb** [comment]
Really?

The GitHub app allowed me to copy this share link:

https://github.com/MarlinFirmware/Marlin/pull/21142#discussion_r839077932

Is that visible?

**Comment by thinkyhead** [comment]
> Is that visible?

If I hover I can see the first few lines of the comment, but the link just highlights itself and doesn't open anything else. I'll try a different browser and see if I get the same result.

**Edit: Same thing in Chrome. The link does nothing.**

**Comment by RFBomb** [comment]
> The link does nothing.

Very odd. 

Clicking the link for me in FIrefox scrolled the page up, and expanded one of the segments, and allowed me to produce this screenshot, which was something I did not expect the link to do.

  <details>
<summary>screenshot</summary>

  </details>


What makes it more interesting is that for a few days they were visible, but currently they are not for me as well. 
Maybe it has to do with the additional commits pushed onto those files? 
After some digging, I did find them. The comments appear to have been marked on the individual commits. 


[Attempt 10 at the link](https://github.com/MarlinFirmware/Marlin/pull/21142/commits/4642925c87e4ec01b16413b86d057bc994f3ba04#diff-ee9800782c3e0dc76daef8f88127ae001599bcf0d552dbf701c07634fcca3cad)

  <details>
<summary>screenshot</summary>

  </details>


[Link that works (finally)](https://github.com/MarlinFirmware/Marlin/pull/21142/commits/971bac725a53855282466dd262127878a70d6865)
  <details>
<summary>screenshot</summary>

  </details>


I believe all of those can be closed out now, as this commit: https://github.com/MarlinFirmware/Marlin/pull/21142/commits/7e30a7855b3417932c78b052b674859731ab7b5b has the changes I had requested for the hover descriptions put back in.


edit: I feel like i'm taking crazy pills. github keeps auto-updating the links, so they point to the commit itself instead of the comment.

edit edit: I was able to finally get it to properly link where I wanted by defining it as a link explicitly, that way github didn't auto-truncate it to point to the wrong page.

File: Marlin/src/module/tool_change.h (hunk 4 above)
Comment (RFBomb) [reply Marlin/src/module/tool_change.h]: Resolved by 👍 https://github.com/MarlinFirmware/Marlin/pull/21142/commits/7e30a7855b3417932c78b052b674859731ab7b5b

**Comment by RFBomb** [comment]
Ok, the links in my above comment are now (finally) working. What a pain that was. 
Though pasting the link to a browser where I am not signed into github results in not being able to view the comments I showed in the screenshots. What a nightmare

**Review by RFBomb — COMMENTED** [review]
My other review comments can be ignored, as they have been resolved or thought about long enough to determine that they are not an issue. 
Maybe this comment will be seen. Maybe my problem was I never hit this 'submit review' button. I dont know, this is the first time trying to use this feature.

File: Marlin/src/module/tool_change.cpp [hunk 5]
```diff
@@ -1094,7 +1090,7 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
             unscaled_e_move(toolchange_settings.extra_prime, MMM_TO_MMS(toolchange_settings.prime_speed));
 
             // Log that this extruder has been primed
-            extruder_primed[new_tool] = true;
+            SBI(extruder_primed, new_tool);
```
[the remark this answers is not in the document]
Comment (RFBomb) [reply Marlin/src/module/tool_change.cpp]: This works, I just wasnt sure how. Ignore

File: Marlin/src/module/tool_change.cpp (hunk 1 above)
Comment (RFBomb) [reply Marlin/src/module/tool_change.cpp]: this is outdated, and has since been confirmed working. Disregard.

**Comment by RFBomb** [comment]
Ah, I never 'submitted' them. I didn't realize that was a thing that had to be done. I assumed it was immediate. This also explains why no date had existed on the comments, and I could also only see them while logged in. 

I marked them all as resolved already @thinkyhead.

**Comment by thinkyhead** [comment]
Thanks! I'm just toying with a more generalized bit-flags thing at the moment, to see how that looks. Marlin aims to save SRAM when possible, but defining a type allows for using an array of `bool` on MCUs where that makes more sense.

File: Marlin/src/module/tool_change.cpp [hunk 6]
```diff
@@ -1051,14 +1162,17 @@ void tool_change(const uint8_t new_tool, bool no_move/*=false*/) {
...
+    // First tool priming. To prime again, reboot the machine. -- Should only occur for first T0 after powerup!
     #if ENABLED(TOOLCHANGE_FS_PRIME_FIRST_USED)
       static bool first_tool_is_primed = false;
-      if (new_tool == old_tool && !first_tool_is_primed && enable_first_prime) {
-        tool_change_prime();
+      
+      if (enable_first_prime && !first_tool_is_primed && new_tool == old_tool && new_tool == 0) {
+        if (!TEST(extruder_was_primed, old_tool)) {
+          tool_change_prime(); //Only perform this prime if it hasn't been primed yet!
+        }
```
Comment (RFBomb) [inline Marlin/src/module/tool_change.cpp]: If you are looking for minor optimizations to save bits, the `first_tool_is_primed` bool can potentially be removed. 

Currently it's statically defined and only set once, and only to ensure this isn't processed again. 

The 'extruder_was_primed' test ( against extruder 0) on the following line can be used in place of it. Not sure how taxing a call like that is compared to checking a bool on its own.

**Comment by bob-anthony** [comment]
I'm sure glad that I found this pull request this morning, as I'd planned on spending most of the day working on the tool change code. The tool change procedure has been the source of much frustration with my CRX. Thank you both for putting in the work to improve this!

An improvement that might be considered is to move the ensure_safe_temperature function from pause.cpp to tempurature.cpp and then utilize it in tool_change.cpp in lieu of too_cold.

**Comment by thinkyhead** [comment]
The `ensure_safe_temperature` function exists to support filament change, where reheating of the hotend is expected or asserted, but in general, it seems safer If we only turn on heaters in response to commands that are closely related to heating. That is more likely to make the user aware that heating is happening and keep them from walking away from the machine thinking it is cold. If we do decide to provide greater 'convenience' and turn on heaters instead of punting that back to the user then we should establish some audible safety warning to alert hearing-capable users that heaters are being turned on.
